src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs

```diff
@@ -22,77 +24,89 @@ namespace Raven.Server.Documents.Sharding.Handlers
+                    AddReplicationPerformance(stats);
+
+                    using (var scope = stats.CreateScope())
                     {
-                        var stats = _lastStats = new OutgoingReplicationStatsAggregator(_parent.GetNextReplicationStatsId(), _lastStats);
-                        AddReplicationPerformance(stats);
+                        EnsureValidStats(scope);
 
-                        using (var scope = stats.CreateScope())
+                        if (hasBatch == false)
                         {
-                            EnsureValidStats(scope);
+                            _lastSentDocumentEtag = _lastEtag;
+                            _lastDocumentSentTime = DateTime.UtcNow;
 
-                            if (items.Count == 0)
-                            {
-                                SendHeartbeat(null);
-                                _replicationQueue.SendToShardCompletion.Signal();
-                                continue;
-                            }
+                            SendHeartbeat(LastDatabaseChangeVector);
+                            continue;
+                        }
 
+                        var batch = _batch;
+                        var tcs = _tcs;
+                        MissingAttachmentsInLastBatch = false;
+
+                        try
+                        {
                             using (_stats.Network.Start())
                             {
-                                MissingAttachmentsInLastBatch = false;
-
-                                var didWork = SendDocumentsBatch(context, items, _stats.Network);
-
-                                if (MissingAttachmentsInLastBatch)
+                                if (batch.Items.Count > 0)
                                 {
-                                    DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Shiran, DevelopmentHelper.Severity.Normal, "Handle missing attachments");
-                                    continue;
+                                    _lastEtag = batch.Items.Last().Etag;
                                 }
-
-                                _replicationQueue.SendToShardCompletion.Signal();
-
-                                if (didWork == false)
-                                    break;
+                                
+                                var didWork = SendDocumentsBatch(context, batch, _stats.Network);
+                                _tcpConnectionOptions._lastEtagSent = _lastEtag;
+                                tcs.TrySetResult();
                             }
                         }
+                        catch (Exception e)
+                        {
+                            tcs.TrySetException(e);
+                            return;
+                            //TODO: is the connection still alive, need to abort?
```

Comment:
yes, you need to ensure the connection is aborted

Reply:
Done